### PR TITLE
CBG-2938: Ignore Cbgt EOF feed errors when intentionally stopped

### PR DESCRIPTION
**Pull request by bbrks**
CBG-2938

- Changed `CbgtContext.StopHeartbeatListener` to `CbgtContext.Stop`
- Made event handler context cancellable from `Stop`
- Made `OnFeedError` callback check context for cancellation before logging about EOF

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1776/
  - `TestResyncUsingDCPStreamForNamedCollection` flake (unrelated)

### CODE REVIEW

File: db/import_listener.go [hunk 1]
```diff
@@ -214,7 +214,7 @@ func (il *importListener) ImportFeedEvent(event sgbucket.FeedEvent) {
 func (il *importListener) Stop() {
 	if il != nil {
 		if il.cbgtContext != nil {
-			il.cbgtContext.StopHeartbeatListener()
+			il.cbgtContext.Stop()
```
Comment (torcolvin) [inline db/import_listener.go:217]: does Stop() stop both?

File: base/dcp_sharded.go [hunk 2]
```diff
@@ -741,8 +748,11 @@ func (meh *sgMgrEventHandlers) OnFeedError(srcType string, r cbgt.Feed, err erro
 		// the connectivity problems either during the next rebalance
 		// (new kv node after failover-recovery rebalance) or
 		// on the next janitor work cycle(ephemeral network issue to the same node).
-		if strings.Contains(err.Error(), "EOF") {
-			InfofCtx(meh.ctx, KeyDCP, "Handling EOF on cbgt feed - notifying manager to trigger reconnection to feed.  indexUUID: %v, err: %v", indexUUID, err)
+		if strings.Contains(feedErr.Error(), "EOF") {
+			// If this wasn't an intentional close, log about the EOF
+			if meh.ctx.Err() != context.Canceled {
```
Comment (torcolvin) [inline base/dcp_sharded.go:769]: slightly prefer errors.Is in case that error gets wrapped

**Review by torcolvin — COMMENTED** [review]
LGTM otherwise

File: db/import_listener.go (hunk 1 above)
Comment (bbrks) [reply db/import_listener.go:217]: yep

File: base/dcp_sharded.go (hunk 2 above)
Comment (bbrks) [reply base/dcp_sharded.go:769]: I'm ok switching this to `!errors.Is(meh.ctx.Err(), context.Canceled)` as it does still work, but errors from `context.Err()` _can't_ be wrapped.

`context` (in 1.20+) has its own way of "wrapping" that still results in `Err()` only returning one of:
- `nil`
- `context.Canceled`
- `context.DeadlineExceeded`

And `errors.Is` still doesn't really support `Err()` as you'd expect it to.

example:
https://go.dev/play/p/XE9rdrP7F9A

Comment (torcolvin) [reply base/dcp_sharded.go:769]: This to leave as is fine, it was just a automatic reaction to seeing an error comparison.

File: base/dcp_sharded.go [hunk 3]
```diff
@@ -439,8 +442,11 @@ func getMinNodeVersion(cfg cbgt.Cfg) (*ComparableVersion, error) {
 	return minVersion, nil
 }
 
-// StopHeartbeatListener unregisters the listener from the heartbeater, and stops it.
-func (c *CbgtContext) StopHeartbeatListener() {
+// Stop unregisters the listener from the heartbeater, and stops it and associated handlers.
+func (c *CbgtContext) Stop() {
```
Comment (adamcfraser) [inline base/dcp_sharded.go:450]: I agree it makes sense to have a more generic CbgtContext.Stop(), but if we're doing that, the rest of the 'on stop' code from importListener.Stop should get moved there as well.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**